# Perform DEG analysis

Load the Libraries

```{r}
library(DESeq2)
library(biomaRt)
library(rtracklayer)
library(GenomicFeatures)
library(ggplot2)
library(dplyr)
library(stringr)
library(ggrepel)
```

Define the conditions

```{r}
conditionOne <- "larva"
conditionTwo <- "pupa"
conditionThree <- "adult"
```

Read in sample descriptions

```{r}
targets <- read.table("targets.txt", header=TRUE, sep="\t")
```

Read in featureCounts for all samples

```{r}
rawData <- NULL

for (sampleID in targets$sampleID) {
    
    fileContents <- read.table(paste("../input/", sampleID, ".markdup.featurecount", sep=""), sep="\t", header=T)
    
    rawData <- cbind(rawData, fileContents[,7])
    
}
```

Add column and row names to the rawData data.frame

```{r}
colnames(rawData) <- targets$sampleID
rownames(rawData) <- fileContents$Geneid
```

Create a DESeq2 design matrix

```{r}
exptDesign = data.frame(

row.names = colnames(rawData),
condition = targets$developmental_stage

)
```

Create a DESeq2 experimental object

```{r}
exptObject <- DESeqDataSetFromMatrix(

countData = rawData,
colData = exptDesign,
design = ~ condition

)
```

Perform DE analysis using DESeq function

```{r}
analysisObject = DESeq(exptObject)
```

Filtering raw data to keep only rows that have at least 18 reads total (because we have 9 samples in total, make sure the gene for each sample has at least 2 reads mapped to it, 9X2=18)

```{r}
keep <- rowSums(counts(analysisObject)) >= 18 
analysisObject <- analysisObject[keep,]
```

# Check the quality of the DEG results by drawing PCA plot, estimating size factors, drawing dispersion plot & building MA plot

1.  Create PCA plot for all the genes (plotting PCA using all genes gives an unbiased view of global expression patterns and sample clustering,)(include as main figure for publication)

```{r}
vsd <- vst(analysisObject, blind = FALSE)
plotPCA(vsd, intgroup = "developmental_stage")
```

Create PCA plot to include only significantly expressed genes (where padj\<0.05)(plotting PCA by using only DEGs highlights variation driven by statistically significant genes but introduces selection bias.)(include as supplementary figure for publication)

```{r}
# Filter results to only include significant genes (padj < 0.05)
res0.05 <- results(analysisObject, alpha = 0.05)
res0.05.df <- as.data.frame(res0.05)

# Subset the significant genes based on padj < 0.05
significant_genes <- rownames(res0.05.df[res0.05.df$padj < 0.05, ])

# Get the normalized counts for significant genes
vsd <- vst(analysisObject, blind = FALSE)
vsd_significant <- vsd[significant_genes, ] # Subset the vsd object to include only significant genes

# Plot PCA for only significant genes
plotPCA(vsd_significant, intgroup = "developmental_stage")
```

2.  Estimate size factors to investigate whether there is any systematic bias in the sequencing data & check for the presence of genes that are less deeply sequenced

```{r}
sizeFactors(analysisObject)
```

3.  Estimate dispersion of data by drawing dispersion plot

```{r}
plotDispEsts(analysisObject)
```

4.  building mean average (MA) plot

```{r}
plotMA(res0.05)
```

Select the top 10 differentially expressed genes

```{r}
library(dplyr)
best_genes <- res0.05.df %>%
  arrange(padj) %>%
  head(10)
write.csv(best_genes, "best_genes.csv")
```

Summarize the DEG results using the summary function

```{r}
res <- results(analysisObject)
summary(res)
```

To display all the genes with p-values less than 0.05

```{r}
res0.05 <- results(analysisObject, alpha = 0.05)
res0.05.df <- as.data.frame(res0.05) # convert to dataframe
str(res0.05.df)
summary(res0.05.df)
```

How many adjusted p-values were less than 0.05 ?

```{r}
sum(res$padj < 0.05, na.rm=TRUE)
```

To display all the genes with p-values less than 0.01

```{r}
res0.01 <- results(analysisObject, alpha = 0.01)
res0.01.df <- as.data.frame(res0.01) 
str(res0.01.df)
summary(res0.01.df)
```

How many adjusted p-values were less than 0.01 ?

```{r}
sum(res$padj < 0.01, na.rm=TRUE)
```

## DESEq2 lets you do custom contrasts between levels of a factor (to compare between larvae & pupa, between larva & adult, and between pupa & adult)

Comparing larva vs pupa (Condition 1 vs Condition 2)

```{r}
res_larva_pupa <- results(analysisObject, contrast=c("condition",conditionOne,conditionTwo))
```

Comparing larva vs adult (Condition 1 vs Condition 3)

```{r}
res_larva_adult <- results(analysisObject,
contrast=c("condition", conditionOne, conditionThree))
```

Comparing pupa vs adult (Condition 2 vs Condition 3)

```{r}
res_pupa_adult <- results(analysisObject,
contrast=c("condition", conditionTwo, conditionThree))
```

Ordering our results table (larva vs pupa for example) by the smallest p value

```{r}
resOrdered <- res_larva_pupa[order(res_larva_pupa$pvalue),]
```

## Find the most upregulated gene per developmental growth stage

Sort the result tables to find the gene with the largest log2FoldChange. Then, intersect top genes from both comparisons to find larva-specific ones

```{r}
# most upregulated gene in larva vs pupa
head(res_larva_pupa[order(res_larva_pupa$log2FoldChange, decreasing=TRUE), ])
```

```{r}
# most upregulated gene in larva vs adult
head(res_larva_adult[order(res_larva_adult$log2FoldChange, decreasing=TRUE), ])
```

Create column for gene names

```{r}
resOrdered$gene_id <- rownames(resOrdered)
```

Reorder the resOrdered object to make the gene_id column as first column

```{r}
resOrdered1 <- resOrdered[,c(7,1:6)]
```

Load the GTF file and extract the gene_id, gene_name and transcript_id columns. Merge the gene_name column with the gene_id column.

```{r}
gtf_file_hs <- import("../resources/Homo_sapiens.GRCh38.109.gtf")
gtf_file_hs1 <- mcols(gtf_file_hs)[,c("gene_id","gene_name")]
resOrdered2 <- data.frame(resOrdered1)
gtf_file_hs2 <- data.frame(gtf_file_hs1)
resOrdered3 <- merge(resOrdered2, gtf_file_hs2, by="gene_id", all.x=T, sort=F)
resOrdered4 <- resOrdered3[,c(1,8,2:7)]
resOrdered_unique <- unique(resOrdered4)
```

Saving the result file as table

```{r}
write.table(resOrdered_unique, quote = F, row.names = F, sep="\t", "../output/larva_vs_pupa_vs_adult.txt")
```

Alternatively, you can also save the result file in .csv format

```{r}
write.csv(resOrdered_unique, "../output/larva_vs_pupa_vs_adult.csv")
```

# Build volcano plot

```{r}
vol <- res0.05.df %>%
  filter(!is.na(padj))
library(ggplot2)
ggplot(vol, aes(x = log2FoldChange,  y = -log10(padj)), color = padj<0.05 & abs(log2FoldChange)>1) + 
  geom_point() +
  geom_text(data = best_genes, aes(label = ensembleGeneID)) +
  geom_text_repel(aes(label = ensembleGeneID), size = 3)
```

# Build heatmap

Read ComplexHeatmap book at <https://jokergoo.github.io/ComplexHeatmap-reference/book/>

Get ComplexHeatmap Rmarkdown script at <https://github.com/brandonyph/ComplexHeatmap_Walkthrough>

```{r}
# Step 1: Select top 50 most significant DEGs
top50_genes <- head(rownames(res0.05.df[order(res0.05.df$padj),]), 50)

# Step 2: Extract normalized expression for those DEGs
vsd_mat <- assay(vsd)[top50_genes, ]

# Step 3: Scale each gene (row-wise z-score)
vsd_mat_scaled <- t(scale(t(vsd_mat)))  # Rows: genes; Columns: samples

# Step 4: Prepare column annotation for developmental stages
# Ensure sample order in targets matches the columns in vsd_mat
targets <- targets[match(colnames(vsd_mat_scaled), targets$sampleID), ]

# Create annotation
col_ha <- HeatmapAnnotation(
  stage = targets$developmental_stage,
  col = list(stage = c(
    larva = "#1f77b4",  # blue
    pupa = "#ff7f0e",   # orange
    adult = "#2ca02c"   # green
  )),
  annotation_legend_param = list(stage = list(title = "Developmental Stage"))
)

# Step 5: Generate the heatmap
Heatmap(vsd_mat_scaled,
        name = "z-score",
        top_annotation = col_ha,
        show_row_names = TRUE,
        show_column_names = TRUE,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        column_split = targets$developmental_stage, # Split columns by larva, pupa, adult
        row_names_gp = gpar(fontsize = 6),
        column_names_gp = gpar(fontsize = 8),
        col = colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
)

```

# Perform orthology-based functional annotation of novel *C. cramerella* genome sequence using eggNOG-mapper

This is done in Linux (refer back to logbook.py)

# Perform annotation mapping files (Extract GO and KEGG terms)

```{r}

library(tidyverse)

# Read the eggNOG annotation file (remove header lines starting with "#")
eggnog <- read_tsv("cramerella_annot.emapper.annotations", comment = "#", col_names = TRUE)

# Preview columns
glimpse(eggnog)

# Extract GO annotations
go_map <- eggnog %>%
  filter(!is.na(GOs)) %>%
  select(query_name, GOs) %>%
  separate_rows(GOs, sep = ",") %>%
  rename(GeneID = query_name, GO_ID = GOs)

# Extract KEGG KO terms
kegg_map <- eggnog %>%
  filter(!is.na(KEGG_ko)) %>%
  select(query_name, KEGG_ko) %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  rename(GeneID = query_name, KEGG_ID = KEGG_ko)

```

# Load DEG gene list

```{r}

# List of significant DEGs (based on adjusted p-value)
sig_genes <- res_annotated %>%
  filter(padj < 0.05) %>%
  pull(ensembl_gene_id) %>%
  unique()

```

# Perform gene set enrichment analysis (GSEA), which includes GO enrichment and KEGG enrichment respectively by using clusterProfiler::enricher()

# GO enrichment

```{r}

library(clusterProfiler)

ego <- enricher(gene = sig_genes,
                TERM2GENE = go_map,
                pvalueCutoff = 0.05,
                pAdjustMethod = "BH")

dotplot(ego, showCategory = 20, font.size = 10)

```

# KEGG enrichment

```{r}

ekegg <- enricher(gene = sig_genes,
                  TERM2GENE = kegg_map,
                  pvalueCutoff = 0.05,
                  pAdjustMethod = "BH")

dotplot(ekegg, showCategory = 20, font.size = 10)

```

# Save the previously extracted GO and KEGG annotation files

```{r}

# Save GO mapping and KEGG mapping
write.csv(go_map, "GO_mapping.csv", row.names = FALSE)
write.csv(kegg_map, "KEGG_mapping.csv", row.names = FALSE)

```
